Guard against instance of GenomeIndexToolData

jdavcs opened the pull request:
The value of `special` can be tracked to [this function definition](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/__init__.py#L1144). The result of the function call may be an instance of `GenomeIndexToolData`, which, unlike `JobExportHistoryArchive`, does *not* have an `fda` attribute. This change guards against such a case which would raise an AttributeError.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

jdavcs committed:
Guard against instance of GenomeIndexToolData

mvdbeek commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -85,7 +85,7 @@ def validate_inputs(input, value, context, **kwargs):
             special = get_special()
             if special:
-                out_data["output_file"] = special.fda
+                out_data["output_file"] = getattr(special, 'fda', None)
```
```suggestion
                out_data["output_file"] = getattr(special, 'fda', fda)
```
`output_file` should be similar to a `HistoryDatasetAssociation` and have a dataset attribute that points to an actual Dataset instance.

Since `get_special` is only called in this one place, could we just return the right thing in `get_special` ?

jdavcs replied:
So, does `get_special` mean this: "return an instance of FakeDatasetAssociation using the appropriate dataset or return None"? And the dataset for the fda would be either provided by JobExportHistoryArchive (inside the `fda` property) or by GenomeIndexToolData (which has a dataset)? If so, maybe we should rename it to `get_fake_dataset_association()`?
(there is [similar logic here](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/__init__.py#L2014), but it's only partially the same, and it does not include GenomeIndexToolData)

jdavcs replied:
i.e.,
```
def get_special_dataset_association():
    # try getting a JobExportHistoryArchive; return its fda
    # if not, try getting a GenomeIndexToolData; return its fda (which I"ll add as a property to that class)
    # if not, return none
```
And then, in tools/evaluation:
```
...
if special:
    out_data["output_file"] = special
```

mvdbeek replied:
Does https://github.com/galaxyproject/galaxy/commit/578fb56533b6e2d6c2aecc1f2699d78ca7648db2 makes sense ? Feel free to cherry-pick that commit if you agree.

mvdbeek replied:
Sorry, got the letters mixed up, https://github.com/galaxyproject/galaxy/commit/7991ee0f3d is better.

jdavcs replied:
Yep, and the commit description is very helpful (I think I understand the logic now - thanks!). Will cherry-pick now.

jdavcs replied:
Done, and fixed the cheery-picking/authorship (Apparently, when I use the `-n` no commit option, it creates a new commit instead of the original. Obviously.)

mvdbeek committed:
Simplify and fix JobExportHistoryArchive / GenomeIndexToolData handling

Output parameters must behave somewhat like a HDA, in that they need to
have a dataset attribute that points to a Dataset instance.
`GenomeIndexToolData` fits that requirement, `JobExportHistoryArchive`
provides the fda property that behaves like a HDA.
